chainstore.go

```diff
@@ -1,147 +1,231 @@
-	for _, s := range c.stores {
-		err = s.Open()
-		if err != nil {
-			return // return first error that comes up
-		}
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	return
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, c *Chain, i int) {
+			s := c.stores[i]
+
+			s.setErr(s.Open())
+
+			wg.Done()
+		}(&wg, c, i)
+	}
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Close() (err error) {
-	for _, s := range c.stores {
-		err = s.Close()
-		// TODO: we shouldn't stop on first error.. should keep trying to close
-		// and record errors separately
-		if err != nil {
-			return
-		}
+// Close closes all the stores.
+func (c *Chain) Close() error {
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Close failed due to a previous error: %q", err)
+	}
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, s *storeWrapper) {
+			s.setErr(s.Close())
+			wg.Done()
+		}(&wg, c.stores[i])
 	}
-	return
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Put(key string, val []byte) (err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Put(ctx context.Context, key string, val []byte) (err error) {
+	if !isValidKey(key) {
 		return ErrInvalidKey
 	}
 
-	fn := func() (err error) {
-		for _, s := range c.stores {
-			err = s.Put(key, val)
-			if err != nil {
-				return
-			}
-		}
-		return
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	if c.async {
-		go fn()
-	} else {
-		err = fn()
+
+	fn := func(s Store) error {
+		return s.Put(ctx, key, val)
 	}
-	return
+
+	return c.doWithContext(ctx, fn)
 }
 
-func (c *Chain) Get(key string) (val []byte, err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Get(ctx context.Context, key string) (val []byte, err error) {
+	if !isValidKey(key) {
 		return nil, ErrInvalidKey
 	}
 
-	for i, s := range c.stores {
-		val, err = s.Get(key)
-		if err != nil {
-			return
-		}
+	if err := c.firstErr(); err != nil {
+		return nil, fmt.Errorf("Open failed due to a previous error: %q", err)
+	}
 
-		if len(val) > 0 {
-			if i > 0 {
-				// put the value in all other stores up the chain
-				fn := func() {
-					for n := i - 1; n >= 0; n-- {
-						c.stores[n].Put(key, val) // errors..?
-					}
-				}
-				// if c.async { } else { } ....?
-				go fn()
+	errCh := make(chan error, 1)
+
+	go func() {
+		for i := range c.stores {
+			s := c.stores[i]
+
+			val, err = s.Get(ctx, key)
+			if err != nil {
+				errCh <- err
+				return
 			}
 
-			// return the first value found on the chain
-			return
+			if val != nil {
+				for j := i - 1; j >= 0; j++ {
+					go c.stores[i].Put(ctx, key, val)
+				}
+				errCh <- nil
+				return
+			}
 		}
+		errCh <- ErrNoSuchKey
+	}()
+
+	select {
+	case <-ctx.Done():
+		c.Close() // Close should unlock pending requests.
```

Comment:
btw: underlying I/O reads should use something like ctxhttp https://godoc.org/golang.org/x/net/context/ctxhttp

yea, we should close only stores that are failing, not because of ctx.Done()  -- and maybe try to reopen them in a while